File: javatests/arcs/core/storage/RamDiskStoreIntegrationTest.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

import arcs.core.crdt.CrdtCount
import arcs.core.crdt.CrdtCount.Operation.Increment
import arcs.core.crdt.CrdtCount.Operation.MultiIncrement
import arcs.core.data.CountType
import arcs.core.storage.ProxyMessage.ModelUpdate
import arcs.core.storage.ProxyMessage.Operations
import arcs.core.storage.driver.RamDisk
import arcs.core.storage.driver.RamDiskDriverProvider
import arcs.core.storage.driver.volatiles.VolatileEntry
import arcs.core.storage.keys.RamDiskStorageKey
import arcs.core.storage.testutil.testDriverFactory
import arcs.core.storage.testutil.testWriteBackProvider
import com.google.common.truth.Truth.assertThat
import kotlin.random.Random
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runBlockingTest
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

/** Tests behaviors of the combination of [DirectStore] and [RamDisk]. */
@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(JUnit4::class)
class RamDiskStoreIntegrationTest {

  @Before
  fun setup() {
    DefaultDriverFactory.update(RamDiskDriverProvider())
    runBlocking {
      RamDisk.clear()
    }
  }

  @Test
  fun stores_sequenceOfModelAndOperationUpdates_asModels() = runBlockingTest {
    val storageKey = RamDiskStorageKey("unique")
    val activeStore = createStore(storageKey)

    val count = CrdtCount()
    count.applyOperation(MultiIncrement(actor = "me", version = 0 to 27, delta = 42))

    // Launch three separate coroutines to emulate some concurrency.
    val jobs = mutableListOf<Job>()
    jobs += launch {
      activeStore.onProxyMessage(ModelUpdate(count.data, id = 1))
      println("Sent 1")
    }
    jobs += launch {
      activeStore.onProxyMessage(
        Operations(listOf(Increment("me", version = 27 to 28)), id = 1)
      )
      println("Sent 2")
    }
    jobs += launch {
      activeStore.onProxyMessage(
        Operations(listOf(Increment("them", version = 0 to 1)), id = 1)
      )
      println("Sent 3")
    }
    println("Jobs length: ${jobs.size}")
    jobs.joinAll()
    println("Joined")

    activeStore.idle()

    val volatileEntry = RamDisk.memory.get<CrdtCount.Data>(storageKey)
    assertThat(volatileEntry?.data).isEqualTo(activeStore.getLocalData())
    assertThat(volatileEntry?.version).isEqualTo(3)
    println("Done")
  }

  @Test
  fun stores_operationUpdates_fromMultipleSources() = runBlockingTest {
    val storageKey = RamDiskStorageKey("unique")
    val activeStore1 = createStore(storageKey)
    val activeStore2 = createStore(storageKey)

    val count1 = CrdtCount()
    count1.applyOperation(MultiIncrement("me", version = 0 to 27, delta = 42))

    val count2 = CrdtCount()
    count2.applyOperation(MultiIncrement("them", version = 0 to 15, delta = 23))

    // These three opearations cannot occur concurrently (if they did, versions wouldn't line up
    // correctly on an actor-by-actor basis, and thus - CRDT would be unable to apply changes.
    coroutineScope {
      async { activeStore1.onProxyMessage(ModelUpdate(count1.data, 1)) }
      async { activeStore2.onProxyMessage(ModelUpdate(count2.data, 1)) }
      async {
        delay(Random.nextLong(1500))
        activeStore1.onProxyMessage(
          Operations(
            listOf(
              Increment("me", version = 27 to 28),
              Increment("other", version = 0 to 1)
            ),
            id = 1
          )
        )
      }
    }

    // These two operations can occur concurrently, since their modifications come from only one
    // actor at a time, and their actors' versions are correct.
    coroutineScope {
      async(start = CoroutineStart.UNDISPATCHED) {
        // Random sleep/delay, to make the ordering of execution random.
        delay(Random.nextLong(1500))
        activeStore2.onProxyMessage(
          Operations(
            listOf(
              Increment("them", version = 15 to 16)
            ),
            id = 1
          )
        )
      }
      async(start = CoroutineStart.UNDISPATCHED) {
        // Random sleep/delay, to make the ordering of execution random.
        delay(Random.nextLong(1500))
        activeStore1.onProxyMessage(
          Operations(
            listOf(
              MultiIncrement("me", version = 28 to 33, delta = 74)
            ),
            id = 1
          )
        )
      }
    }

    activeStore1.idle()
    activeStore2.idle()

    val volatileEntry: VolatileEntry<CrdtCount.Data>? = RamDisk.memory.get(storageKey)
    assertThat(volatileEntry?.data).isEqualTo(activeStore1.getLocalData())
    assertThat(volatileEntry?.data).isEqualTo(activeStore2.getLocalData())
    assertThat(volatileEntry?.version).isEqualTo(5)
  }

  @Test
  @Suppress("UNCHECKED_CAST")
  fun store_operationUpdates_fromMultipleSources_withTimingDelays() = runBlockingTest {
    val storageKey = RamDiskStorageKey("unique")
    val activeStore1 = createStore(storageKey)
    val activeStore2 = createStore(storageKey)

    activeStore1.onProxyMessage(Operations(listOf(Increment("me", 0 to 1)), 1))

    delay(100)

    val concurrentJobA = launch {
      delay(Random.nextLong(5))
      activeStore1.onProxyMessage(
        Operations(listOf(Increment("them", 0 to 1)), 1)
      )
    }
    val concurrentJobB = launch {
      delay(Random.nextLong(5))
      activeStore2.onProxyMessage(
        Operations(listOf(Increment("other", 0 to 1)), 1)
      )
    }

    listOf(concurrentJobA, concurrentJobB).joinAll()

    delay(100)

    activeStore2.onProxyMessage(Operations(listOf(Increment("other", 1 to 2)), 1))

    activeStore1.idle()
    activeStore2.idle()

    val entry: VolatileEntry<CrdtCount.Data>? = RamDisk.memory.get(storageKey)
    assertThat(entry?.data).isEqualTo(activeStore1.getLocalData())
    assertThat(entry?.data).isEqualTo(activeStore2.getLocalData())
    assertThat(entry?.version).isEqualTo(4)
    assertThat(activeStore1.localModel.consumerView).isEqualTo(4)
    assertThat(activeStore2.localModel.consumerView).isEqualTo(4)
  }

  companion object {
    private suspend fun CoroutineScope.createStore(
      storageKey: StorageKey
    ): DirectStore<CrdtCount.Data, CrdtCount.Operation, Int> {
      return DirectStore.create(
        StoreOptions(
          storageKey,
          type = CountType()
        ),
        this,
        testDriverFactory,
        ::testWriteBackProvider,
        null
      )
    }
  }
}

File: src/runtime/storage/tests/ramdisk-store-integration-test.ts
/**
 * @license
 * Copyright (c) 2019 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from '../../../platform/chai-web.js';
import {ProxyMessageType} from '../store-interface.js';
import {CRDTCountTypeRecord, CRDTCount, CountOpTypes} from '../../../crdt/lib-crdt.js';
import {RamDiskStorageKey} from '../drivers/ramdisk.js';
import {Exists} from '../drivers/driver.js';
import {Runtime} from '../../runtime.js';
import {CountType} from '../../../types/lib-types.js';
import {StorageKey} from '../storage-key.js';
import {StoreInfo} from '../store-info.js';
import {ActiveStore} from '../active-store.js';
import {DirectStorageEndpointManager} from '../direct-storage-endpoint-manager.js';

async function createStore(storageKey: StorageKey, exists: Exists): Promise<ActiveStore<CRDTCountTypeRecord>> {
  return (await new DirectStorageEndpointManager().getActiveStore(new StoreInfo({
      storageKey, type: new CountType(), exists, id: 'an-id'}))) as ActiveStore<CRDTCountTypeRecord>;
}

describe('RamDisk + Store Integration', async () => {

  afterEach(() => {
    Runtime.resetDrivers();
  });

  it('will store a sequence of model and operation updates as models', async () => {
    const runtime = new Runtime();
    const storageKey = new RamDiskStorageKey('unique');
    const activeStore = await createStore(storageKey, Exists.ShouldCreate);

    const count = new CRDTCount();
    count.applyOperation({type: CountOpTypes.MultiIncrement, actor: 'me', value: 42, version: {from: 0, to: 27}});

    await activeStore.onProxyMessage({type: ProxyMessageType.ModelUpdate, model: count.getData(), id: 1});
    await activeStore.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'me', version: {from: 27, to: 28}}
    ], id: 1});
    await activeStore.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'them', version: {from: 0, to: 1}}
    ], id: 1});

    const volatileEntry = runtime.getMemoryProvider().getVolatileMemory().entries.get(storageKey.unique);
    assert.deepEqual(volatileEntry.root.data, activeStore['localModel'].getData());
    assert.strictEqual(volatileEntry.root.version, 3);
  });

  it('will store operation updates from multiple sources', async () => {
    const runtime = new Runtime();
    const storageKey = new RamDiskStorageKey('unique');
    const activeStore1 = await createStore(storageKey, Exists.ShouldCreate);
    const activeStore2 = await createStore(storageKey, Exists.ShouldExist);

    const count1 = new CRDTCount();
    count1.applyOperation({type: CountOpTypes.MultiIncrement, actor: 'me', value: 42, version: {from: 0, to: 27}});

    const count2 = new CRDTCount();
    count2.applyOperation({type: CountOpTypes.MultiIncrement, actor: 'them', value: 23, version: {from: 0, to: 15}});

    const modelReply1 = activeStore1.onProxyMessage({type: ProxyMessageType.ModelUpdate, model: count1.getData(), id: 1});
    const modelReply2 = activeStore2.onProxyMessage({type: ProxyMessageType.ModelUpdate, model: count2.getData(), id: 1});

    const opReply1 = activeStore1.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'me', version: {from: 27, to: 28}},
      {type: CountOpTypes.Increment, actor: 'other', version: {from: 0, to: 1}}
    ], id: 1});
    const opReply2 = activeStore2.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'them', version: {from: 15, to: 16}},
    ], id: 1});
    const opReply3 = activeStore1.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.MultiIncrement, actor: 'me', value: 74, version: {from: 28, to: 33}},
    ], id: 1});

    await Promise.all([modelReply1, modelReply2, opReply1, opReply2, opReply3]);

    await activeStore1.idle();
    await activeStore2.idle();

    const volatileEntry = runtime.getMemoryProvider().getVolatileMemory().entries.get(storageKey.unique);
    assert.deepEqual(volatileEntry.root.data, activeStore1['localModel'].getData());
    assert.strictEqual(volatileEntry.root.version, 3);
  });

  it('will store operation updates from multiple sources with some timing delays', async () => {
    // store1.onProxyMessage, DELAY, DELAY, DELAY, store1.onProxyMessage, store2.onProxyMessage, DELAY, DELAY, DELAY, store2.onProxyMessage, DELAY, DELAY, DELAY, DELAY, DELAY
    const runtime = new Runtime();
    const storageKey = new RamDiskStorageKey('unique');
    const activeStore1 = await createStore(storageKey, Exists.ShouldCreate);
    const activeStore2 = await createStore(storageKey, Exists.ShouldExist);

    const opReply1 = activeStore1.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'me', version: {from: 0, to: 1}}
    ], id: 1});

    await 0;
    await 0;
    await 0;

    const opReply2 = activeStore1.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'them', version: {from: 0, to: 1}},
    ], id: 1});

    const opReply3 = activeStore2.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'other', version: {from: 0, to: 1}},
    ], id: 1});

    await 0;
    await 0;
    await 0;

    const opReply4 = activeStore2.onProxyMessage({type: ProxyMessageType.Operations, operations: [
      {type: CountOpTypes.Increment, actor: 'other', version: {from: 1, to: 2}},
    ], id: 1});

    await Promise.all([opReply1, opReply2, opReply3, opReply4]);

    await activeStore1.idle();
    await activeStore2.idle();

    const volatileEntry = runtime.getMemoryProvider().getVolatileMemory().entries.get(storageKey.unique);
    assert.deepEqual(volatileEntry.root.data, activeStore1['localModel'].getData());
    assert.strictEqual(volatileEntry.root.version, 4);
  });
});
